File: quoter/src/transaction_builder/order_id.rs
use alloy::primitives::keccak256;

#[derive(Debug, Clone)]
pub struct OrderData {
    pub version: u16,
    pub sender: [u8; 32],
    pub nonce: u64,
    pub origin_chain_id: u32,
    pub dest_chain_id: u32,
    pub fill_deadline: u64,
    pub token_in: [u8; 32],
    pub token_out: [u8; 32],
    pub amount_in: u128,
    pub amount_out: u128,
    pub recipient: [u8; 32],
    pub solver: [u8; 32],
}

impl OrderData {
    /// Encode the order data as bytes for hashing
    pub fn encode(&self) -> Vec<u8> {
        let mut encoded: Vec<u8> = Vec::with_capacity(206);

        encoded.extend_from_slice(&self.version.to_be_bytes());
        encoded.extend_from_slice(&self.sender);
        encoded.extend_from_slice(&self.nonce.to_be_bytes());
        encoded.extend_from_slice(&self.origin_chain_id.to_be_bytes());
        encoded.extend_from_slice(&self.dest_chain_id.to_be_bytes());
        encoded.extend_from_slice(&self.fill_deadline.to_be_bytes());
        encoded.extend_from_slice(&self.token_in);
        encoded.extend_from_slice(&self.token_out);
        encoded.extend_from_slice(&self.amount_in.to_be_bytes());
        encoded.extend_from_slice(&self.amount_out.to_be_bytes());
        encoded.extend_from_slice(&self.recipient);
        encoded.extend_from_slice(&self.solver);

        encoded
    }

    pub fn compute_order_id(&self) -> [u8; 32] {
        keccak256(&self.encode()).0
    }
}

File: svm/scripts/lib/order-id.ts
import { keccak256 } from "js-sha3";

/**
 * OrderData matches the cross-chain canonical encoding from
 * svm/programs/order_book/src/state/orders.rs
 */
export interface OrderData {
  version: number; // u16
  sender: Uint8Array; // 32 bytes
  nonce: bigint; // u64
  originChainId: number; // u32
  destChainId: number; // u32
  createdAt: bigint; // u64
  fillDeadline: bigint; // u64
  tokenIn: Uint8Array; // 32 bytes
  tokenOut: Uint8Array; // 32 bytes
  amountIn: bigint; // u128
  amountOut: bigint; // u128
  recipient: Uint8Array; // 32 bytes
  solver: Uint8Array; // 32 bytes
}

// Total encoded size: 2 + 32 + 8 + 4 + 4 + 8 + 8 + 32 + 32 + 16 + 16 + 32 + 32 = 226
const ENCODED_SIZE = 226;

function writeBE(buf: Buffer, offset: number, value: bigint, bytes: number): void {
  for (let i = bytes - 1; i >= 0; i--) {
    buf[offset + i] = Number(value & 0xffn);
    value >>= 8n;
  }
}

/**
 * Encode OrderData into a packed byte array using big-endian encoding.
 * This must match the Rust/EVM encoding exactly.
 */
export function encodeOrderData(order: OrderData): Buffer {
  const buf = Buffer.alloc(ENCODED_SIZE);
  let offset = 0;

  // version: u16 BE
  buf.writeUInt16BE(order.version, offset);
  offset += 2;

  // sender: 32 bytes
  Buffer.from(order.sender).copy(buf, offset);
  offset += 32;

  // nonce: u64 BE
  writeBE(buf, offset, order.nonce, 8);
  offset += 8;

  // originChainId: u32 BE
  buf.writeUInt32BE(order.originChainId, offset);
  offset += 4;

  // destChainId: u32 BE
  buf.writeUInt32BE(order.destChainId, offset);
  offset += 4;

  // createdAt: u64 BE
  writeBE(buf, offset, order.createdAt, 8);
  offset += 8;

  // fillDeadline: u64 BE
  writeBE(buf, offset, order.fillDeadline, 8);
  offset += 8;

  // tokenIn: 32 bytes
  Buffer.from(order.tokenIn).copy(buf, offset);
  offset += 32;

  // tokenOut: 32 bytes
  Buffer.from(order.tokenOut).copy(buf, offset);
  offset += 32;

  // amountIn: u128 BE
  writeBE(buf, offset, order.amountIn, 16);
  offset += 16;

  // amountOut: u128 BE
  writeBE(buf, offset, order.amountOut, 16);
  offset += 16;

  // recipient: 32 bytes
  Buffer.from(order.recipient).copy(buf, offset);
  offset += 32;

  // solver: 32 bytes
  Buffer.from(order.solver).copy(buf, offset);
  offset += 32;

  return buf;
}

/**
 * Compute order ID as keccak256 hash of encoded OrderData.
 * Returns 32-byte Uint8Array.
 */
export function computeOrderId(order: OrderData): Uint8Array {
  const encoded = encodeOrderData(order);
  return new Uint8Array(keccak256.arrayBuffer(encoded));
}
